### ISSUE TITLE
TerarkZipTable causes compile error

### ISSUE THREAD

**Issue by superdolt** [bytedance/terarkdb#104]
i use open_column_families a lot so i thought about using terarkzip as default with appended code below BUT...
if anyone can help resolve this, i may be able to track down 1 bug i think is in terarkdb with column families. pls help

i dunno how to code in c++ actually for a long time. seems like this is c++11 thing.

```
terarkdb/db/c.cc:729:35: error: cannot bind rvalue reference of type 'std::shared_ptr<terarkdb::TableFactory>&&' to lvalue of type 'std::shared_ptr<terarkdb::TableFactory>'
  729 |   db_options->rep.table_factory = table_factory;
      |                                   ^~~~~~~~~~~~~
```

how do i resolve the above, the code is as below...

```
rocksdb_t* rocksdb_open_column_families(
    const rocksdb_options_t* db_options, const char* name,
    int num_column_families, const char** column_family_names,
    const rocksdb_options_t** column_family_options,
    rocksdb_column_family_handle_t** column_family_handles, char** errptr) {
  std::vector<ColumnFamilyDescriptor> column_families;
  for (int i = 0; i < num_column_families; i++) {
    column_families.push_back(ColumnFamilyDescriptor(
        std::string(column_family_names[i]),
        ColumnFamilyOptions(column_family_options[i]->rep)));
  }

  DB* db;
  std::vector<ColumnFamilyHandle*> handles;








  BlockBasedTableOptions blockbased_options;
  blockbased_options.block_size = 8ULL << 10;

  std::shared_ptr<terarkdb::TableFactory> table_factory;
  table_factory.reset(NewBlockBasedTableFactory(blockbased_options));

  terarkdb::TerarkZipTableOptions tzt_options;
  tzt_options.localTempDir = "/tmp/t";
  tzt_options.indexNestLevel = 3;
  tzt_options.sampleRatio = 0.01;
  tzt_options.terarkZipMinLevel = 2; // Start using TerarkZipTable from level 2

  table_factory.reset(NewTerarkZipTableFactory(tzt_options, table_factory));


  db_options->rep.table_factory = table_factory; //crashed here



```

**Comment by skyzh** [comment]
It seems that you directly changed the source code of `rocksdb_open_column_families` to force enable `TerarkZipTable`. This is not recommended. You may refer to `db_bench_tool.cc` for how to set a TerarkDB instance with `TerarkZipTable` enabled. Specifically,

```cpp
      options.table_factory.reset(TERARKDB_NAMESPACE::NewTerarkZipTableFactory(
          tzto, options.table_factory));
```

**Comment by superdolt** [comment]
i know it's not recommended BUT i'm using other language bindings and the binding is more difficult to do than just this code modification. this is for testing first. possible to help to force use terark zip? how do i change the code properly to achieve this?

**Comment by skyzh** [comment]
> i know it's not recommended BUT i'm using other language bindings and the binding is more difficult to do than just this code modification. this is for testing first. possible to help to force use terark zip? how do i change the code properly to achieve this?

We do not provide official support for go bindings. You can set a default value for `options.table_factory`, in your case.

**Comment by superdolt** [comment]
@skyzh is it possible for you to let us close the issue ourselves and only after some time, u can close it? like wait 7-14 days coz we cant find the issue we submitted and this seemed like spamming to need to repeat.

1. where do u set the default value for options.table_factory? thx in advance. as with other repos i've used on github, will prefer if i can close this so as to mark resolved for all other users in future.

2. by the way, i've tried running dbbench and it doesnt seem very informative. dont even see sst files being generated.

**Comment by skyzh** [comment]
>  is it possible for you to let us close the issue ourselves and only after some time, u can close it?

People use GitHub issues to report bugs and to request features, not for discussion and tutorials. None of your issues follow the templates. Even though you use GitHub issues for help and discussions, your question 1. has been answered by me 2. is not caused by TerarkDB source code, but because you are using it in the wrong way.

> find the issue we submitted and this seemed like spamming to need to repeat

No one will be able to find what they need if they submit issues as you did.

> where do u set the default value for options.table_factory?

https://github.com/bytedance/terarkdb/issues/107

Just like how you do it with RocksDB. I believe the All-In-One Docs has already covered this scene. Please RTFM. If you still cannot compile this or do not even have a basic understanding of C++, reach us in Slack for tutorials and guides. See README for more information.

```
#include <cassert>
#include "rocksdb/db.h"
#include "rocksdb/options.h"
#include "rocksdb/table.h"
#include "table/terark_zip_table.h"

// inside some function

rocksdb::DB* db;
rocksdb::Options options;
// Your db options here
options.create_if_missing = true;
options.wal_bytes_per_sync = 32768;
options.bytes_per_sync = 32768;
// TerarkZipTable need a `fallback` options because you can indicate which LSM level you want to start using TerarkZipTable
// For example, by setting tzt_options.terarkZipMinLevel = 2, TerarkDB will use your fallback Table on level 0 and 1.
std::shared_ptr<terarkdb::TableFactory> table_factory;
rocksdb::BlockBasedTableOptions blockbased_options;
blockbased_options.block_size = 8ULL << 10;
table_factory.reset(NewBlockBasedTableFactory(blockbased_options));
rocksdb::TerarkZipTableOptions tzt_options;
// TerarkZipTable requires a temp directory other than data directory, a slow device is acceptable
tzt_options.localTempDir = "/tmp";
tzt_options.indexNestLevel = 3;
tzt_options.sampleRatio = 0.01;
tzt_options.terarkZipMinLevel = 2; // Start using TerarkZipTable from level 2
table_factory.reset(rocksdb::NewTerarkZipTableFactory(tzt_options, table_factory));
options.table_factory = table_factory;
// Open DB
auto status = rocksdb::DB::Open(options, "/tmp/testdb2", &db);
// Operations
std::string value;
auto s = db->Put(rocksdb::WriteOptions(), "key1", "value1");
s = db->Get(rocksdb::ReadOptions(), "key1", &value);
      assert(s.ok());
assert("value1" == value);
s = db->Delete(rocksdb::WriteOptions(), "key1");
assert(s.ok());
```

> by the way, i've tried running dbbench and it doesnt seem very informative. dont even see sst files being generated

Please provide a minimum configuration that produces a "doesnt seem very informative" result. In my case, it works perfectly.

---

I'm not providing customer service that makes every user happy. If you cannot provide enough information for us to identify problems, or just using GitHub issues to ask for help without any useful information, I'm afraid I could not help in any form. But in other cases, I'm always willing to help and build a better TerarkDB community.

**Comment by hiqsociety** [comment]
1. one last question, possible to provide instructions where i can modify this code to use terarkzip instead of block based? because to write the binding in golang is too much work just for terarkzip (since i have no idea where to start the binding).

i've managed to get golang working without terarkzip, just terark. compiled and replaced with libterark.a etc which all works but when it comes to trying to get it to work with terarkzip, there seems to be too much work to do the binding. appreciate if you can provide the location of the files where i need to edit to make it default using terarkzip because it is using block based now.

**Comment by superdolt** [comment]
one last question, possible to provide instructions where i can modify this code to use terarkzip instead of block based? because to write the binding in golang is too much work just for terarkzip (since i have no idea where to start the binding).

i've managed to get golang working without terarkzip, just terark. compiled and replaced with libterark.a etc which all works but when it comes to trying to get it to work with terarkzip, there seems to be too much work to do the binding. appreciate if you can provide the location of the files where i need to edit to make it default using terarkzip because it is using block based now.

**Comment by skyzh** [comment]
I believe you could edit default option here.

```cpp
ColumnFamilyOptions::ColumnFamilyOptions()
    : compression(Snappy_Supported() ? kSnappyCompression : kNoCompression),
      table_factory(
          std::shared_ptr<TableFactory>(new BlockBasedTableFactory())) {}
```

https://github.com/bytedance/terarkdb/blob/dev.1.4/options/options.cc#L101

**Comment by superdolt** [comment]
sry not well verse with c++,
is this correct? possible to pls help fix this code? not sure how to do the "callback" within the function

```
  terarkdb::TerarkZipTableOptions tzt_options;
  tzt_options.localTempDir = "/tmp/t";
  tzt_options.indexNestLevel = 3;
  tzt_options.sampleRatio = 0.01;
  tzt_options.terarkZipMinLevel = 2; // Start using TerarkZipTable from level 2

ColumnFamilyOptions::ColumnFamilyOptions()
    : compression(Snappy_Supported() ? kSnappyCompression : kNoCompression),
      table_factory.reset(
          NewTerarkZipTableFactory(tzt_options, table_factory)) {}



//      table_factory(
//          std::shared_ptr<TableFactory>(new BlockBasedTableFactory())) {}

```
